File: backend/src/routes/carteirinhas.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { getDb } = require('../db/schema');
const { autenticar, filtrarCongregacao } = require('../middleware/auth');
const { enfileirar } = require('../services/sync');

const router = express.Router();
router.use(autenticar, filtrarCongregacao);

router.get('/', (req, res) => {
  const db = getDb();
  const filtros = ['1=1'];
  const params = [];
  if (req.congregacaoFiltro) { filtros.push('m.congregacao_id = ?'); params.push(req.congregacaoFiltro); }
  if (req.query.status) { filtros.push('c.status = ?'); params.push(req.query.status); }
  const rows = db.prepare(`
    SELECT c.*, m.nome, m.foto_url, m.cargo, m.congregacao_id,
           cg.nome as congregacao_nome
    FROM carteirinhas c
    JOIN membros m ON c.membro_id = m.id
    JOIN congregacoes cg ON m.congregacao_id = cg.id
    WHERE ${filtros.join(' AND ')}
    ORDER BY m.nome
  `).all(...params);
  res.json(rows);
});

// Lista membros sem carteirinha
router.get('/pendentes', (req, res) => {
  const db = getDb();
  const where = req.congregacaoFiltro ? `AND m.congregacao_id = '${req.congregacaoFiltro}'` : '';
  const rows = db.prepare(`
    SELECT m.id, m.nome, m.foto_url, m.cargo, m.congregacao_id
    FROM membros m
    WHERE m.status = 'ativo' ${where}
    AND m.id NOT IN (SELECT membro_id FROM carteirinhas WHERE status != 'nao_emitida')
    ORDER BY m.nome
  `).all();
  res.json(rows);
});

router.post('/emitir', (req, res) => {
  const db = getDb();
  const { membro_id, obreiro_id, validade_meses = 12 } = req.body;
  if (!membro_id) return res.status(400).json({ erro: 'membro_id obrigatório' });

  const hoje = new Date();
  const validade = new Date(hoje);
  validade.setMonth(validade.getMonth() + parseInt(validade_meses));

  const id = uuidv4();
  const qrcode_hash = Buffer.from(`KAIROS:${membro_id}:${id}`).toString('base64');

  // Desativa carteirinha anterior se existir
  db.prepare("UPDATE carteirinhas SET status = 'vencida' WHERE membro_id = ? AND status = 'ativa'").run(membro_id);

  const dados = {
    id,
    membro_id,
    obreiro_id: obreiro_id || null,
    qrcode_hash,
    emissao: hoje.toISOString().split('T')[0],
    validade: validade.toISOString().split('T')[0],
    status: 'ativa',
    criado_em: hoje.toISOString(),
  };

  db.prepare(`INSERT INTO carteirinhas (id, membro_id, obreiro_id, qrcode_hash, emissao, validade, status, criado_em)
    VALUES (@id, @membro_id, @obreiro_id, @qrcode_hash, @emissao, @validade, @status, @criado_em)`).run(dados);

  enfileirar('carteirinhas', 'upsert', dados);

  // Atualiza vencidas automaticamente
  db.prepare("UPDATE carteirinhas SET status = 'vencida' WHERE date(validade) < date('now') AND status = 'ativa'").run();

  const carteirinha = db.prepare(`
    SELECT c.*, m.nome, m.foto_url, m.cargo, cg.nome as congregacao_nome
    FROM carteirinhas c
    JOIN membros m ON c.membro_id = m.id
    JOIN congregacoes cg ON m.congregacao_id = cg.id
    WHERE c.id = ?
  `).get(id);

  res.status(201).json(carteirinha);
});

router.delete('/:id', (req, res) => {
  const db = getDb();
  db.prepare('DELETE FROM carteirinhas WHERE id = ?').run(req.params.id);
  res.json({ ok: true });
});

module.exports = router;

File: backend/app/routes/carteirinhas.py
import base64
from datetime import date
from typing import Optional
from dateutil.relativedelta import relativedelta
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.deps import congregacao_filter, get_current_user
from app.models import Carteirinha, Congregacao, Membro, Usuario
from app.utils import new_id

router = APIRouter(prefix="/carteirinhas", tags=["carteirinhas"])

class EmitirIn(BaseModel):
    membro_id: str
    obreiro_id: Optional[str] = None
    validade_meses: int = 12

def _serializar(c: Carteirinha, membro: Membro, congregacao: Congregacao) -> dict:
    return {
        "id": c.id, "membro_id": c.membro_id, "obreiro_id": c.obreiro_id, "qrcode_hash": c.qrcode_hash,
        "foto_url": c.foto_url, "emissao": c.emissao, "validade": c.validade, "status": c.status,
        "nome": membro.nome, "cargo": membro.cargo, "congregacao_id": membro.congregacao_id,
        "congregacao_nome": congregacao.nome,
    }

@router.get("")
def listar(status: str = "", db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
           cong_filtro: Optional[str] = Depends(congregacao_filter)):
    q = (
        db.query(Carteirinha, Membro, Congregacao)
        .join(Membro, Carteirinha.membro_id == Membro.id)
        .join(Congregacao, Membro.congregacao_id == Congregacao.id)
        .filter(Carteirinha.tenant_id == cu.tenant_id)
    )
    if cong_filtro:
        q = q.filter(Membro.congregacao_id == cong_filtro)
    if status:
        q = q.filter(Carteirinha.status == status)
    rows = q.order_by(Membro.nome).all()
    return [_serializar(c, m, cg) for c, m, cg in rows]

@router.get("/pendentes")
def pendentes(db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
              cong_filtro: Optional[str] = Depends(congregacao_filter)):
    emitidas_subq = db.query(Carteirinha.membro_id).filter(Carteirinha.status != "nao_emitida")
    q = db.query(Membro).filter(Membro.tenant_id == cu.tenant_id, Membro.status == "ativo",
                                 Membro.id.notin_(emitidas_subq))
    if cong_filtro:
        q = q.filter(Membro.congregacao_id == cong_filtro)
    membros = q.order_by(Membro.nome).all()
    return [{"id": m.id, "nome": m.nome, "foto_url": m.foto_url, "cargo": m.cargo,
             "congregacao_id": m.congregacao_id} for m in membros]

@router.post("/emitir", status_code=201)
def emitir(payload: EmitirIn, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user)):
    membro = db.query(Membro).filter(Membro.id == payload.membro_id, Membro.tenant_id == cu.tenant_id).first()
    if not membro:
        raise HTTPException(status_code=404, detail="Membro não encontrado")

    hoje = date.today()
    validade = hoje + relativedelta(months=payload.validade_meses)

    db.query(Carteirinha).filter(Carteirinha.membro_id == payload.membro_id,
                                  Carteirinha.status == "ativa").update({"status": "vencida"})

    nova_id = new_id()
    qrcode_hash = base64.b64encode(f"KAIROS:{payload.membro_id}:{nova_id}".encode()).decode()
    carteirinha = Carteirinha(
        id=nova_id, tenant_id=cu.tenant_id, membro_id=payload.membro_id, obreiro_id=payload.obreiro_id,
        qrcode_hash=qrcode_hash, emissao=hoje, validade=validade, status="ativa",
    )
    db.add(carteirinha)

    db.query(Carteirinha).filter(Carteirinha.tenant_id == cu.tenant_id, Carteirinha.status == "ativa",
                                  Carteirinha.validade < hoje).update({"status": "vencida"})
    db.commit(); db.refresh(carteirinha)

    congregacao = db.query(Congregacao).filter(Congregacao.id == membro.congregacao_id).first()
    return _serializar(carteirinha, membro, congregacao)

@router.delete("/{carteirinha_id}")
def remover(carteirinha_id: str, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user)):
    c = db.query(Carteirinha).filter(Carteirinha.id == carteirinha_id, Carteirinha.tenant_id == cu.tenant_id).first()
    if c:
        db.delete(c); db.commit()
    return {"ok": True}
